### PR TITLE
repr: add support for all compacted date formats

### PR DESCRIPTION
**Pull request by petrosagg**
Adds support for dates like `700203`, `0010203` and `200102031`

cc @quodlibetor

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4889)

### CODE REVIEW

**Comment by benesch** [comment]
Thanks! Triggered a build for you.

File: src/repr/src/adt/datetime.rs [hunk 1]
```diff
@@ -1362,9 +1376,12 @@ pub(crate) fn tokenize_time_str(value: &str) -> Result<VecDeque<TimeStrToken>, S
     let mut num_buf = String::with_capacity(4);
     let mut char_buf = String::with_capacity(7);
     fn parse_num(n: &str, idx: usize) -> Result<TimeStrToken, String> {
-        Ok(TimeStrToken::Num(n.parse().map_err(|e| {
-            format!("Unable to parse value as a number at index {}: {}", idx, e)
-        })?))
+        Ok(TimeStrToken::Num(
+            n.parse().map_err(|e| {
+                format!("Unable to parse value as a number at index {}: {}", idx, e)
```
Comment (quodlibetor) [inline src/repr/src/adt/datetime.rs:1381]: To make this match our other error messages could you include the value that failed to parse in the message:

```suggestion
                format!("Unable to parse value {} as a number at index {}: {}", n, idx, e)
```

**Review by quodlibetor — COMMENTED** [review]
This looks great!

One small request for the error message, but I like the fact that this also should make our redisplay of time strings more closely adhere to the original value.

File: src/repr/src/adt/datetime.rs (hunk 1 above)
Comment (petrosagg) [reply src/repr/src/adt/datetime.rs]: Fixed, thanks!

**Comment by petrosagg** [comment]
@quodlibetor I fixed the tests for the updated error message but I see that it can produce some slightly confusing errors. For example this input `"-9223372036854775808 seconds"` produces the error `"Unable to parse value 9223372036854775808 as a number at index 20: number too large to fit in target type"` (notice the absence of the minus sign) because it's consumed by the parser at an earlier step as a `Dash`.

What do you think?

**Comment by quodlibetor** [comment]
Hi @petrosagg sorry this somehow fell out of my notifications! I'll keep a closer eye on the PRs page.

> notice the absence of the minus sign because it's consumed by the parser at an earlier step as a Dash.

Gah, I forgot that we capture the raw value earlier in the error case, the tests are failing with a message that is pretty helpful:

```text
invalid input syntax for interval: Unable to parse value 9223372036854775808 as a number at index 20: number too large to fit in target type: "-9223372036854775808 months"
```

sorry about that!

**Comment by petrosagg** [comment]
@quodlibetor no worries! I wasn't sure when was the right time to reping you :). Ok, let me restore the error message and get this merged

**Comment by petrosagg** [comment]
@quodlibetor removed `n` and rebased on top of main. Can you re-trigger the CI?

**Comment by benesch** [comment]
Done.

**Comment by quodlibetor** [comment]
>  I wasn't sure when was the right time to reping you :)

I try to check these at least twice a (business) day, if it's been 24 hours definitely reping!

File: src/repr/tests/strconv.rs [hunk 2]
```diff
@@ -15,6 +15,12 @@ use repr::strconv;
 
 #[test]
 fn test_parse_date() {
+    run_test_parse_date("690203", NaiveDate::from_ymd(2069, 2, 3));
+    run_test_parse_date("700203", NaiveDate::from_ymd(1970, 2, 3));
+    run_test_parse_date("010203", NaiveDate::from_ymd(2001, 2, 3));
+    run_test_parse_date("0010203", NaiveDate::from_ymd(1, 2, 3));
+    run_test_parse_date("00010203", NaiveDate::from_ymd(1, 2, 3));
+    run_test_parse_date("20010203", NaiveDate::from_ymd(2001, 2, 3));
     run_test_parse_date("2001-02-03", NaiveDate::from_ymd(2001, 2, 3));
```
Comment (quodlibetor) [inline src/repr/tests/strconv.rs:24]: could you include a couple of negative cases (e.g. from the earlier draft) as well?

something like `20010199` and `-12345678910111213141516171819`.

**Review by quodlibetor — COMMENTED** [review]
Other than a couple negative tests this looks fantastic!

**Pull request merged**